### PR TITLE
C# 7.x: Adding more support for expression-bodied members

### PR DESCRIPTION
**Pull request by RexJaeschke**

### CODE REVIEW

File: standard/classes.md [hunk 1]
```diff
@@ -3094,7 +3095,10 @@ The use of *accessor_modifier*s is governed by the following restrictions:
...
+For `abstract` and `extern` properties, the *accessor_body* for each accessor specified is simply a semicolon. A non-abstract, non-extern property may also have the *accessor_body* for all accessors specified be a semicolon, in which case it is an ***automatically implemented property*** ([§14.7.4](classes.md#1474-automatically-implemented-properties)). An automatically implemented property shall have at least a get accessor. For the accessors of any other non-abstract, non-extern property, the *accessor_body* is either
+
+- a *block* that specifies the statements to be executed when the corresponding accessor is invoked.
+- or an expression body, which consists of `=>` followed by an *expression* and a semicolon, and denotes a single expression to be executed when the corresponding accessor is invoked.
```
Comment (jskeet) [inline standard/classes.md]:
```suggestion
For `abstract` and `extern` properties, the *accessor_body* for each accessor specified is simply a semicolon. A non-abstract, non-extern property may also have the *accessor_body* for all accessors specified be a semicolon, in which case it is an ***automatically implemented property*** ([§14.7.4](classes.md#1474-automatically-implemented-properties)). An automatically implemented property shall have at least a get accessor. For the accessors of any other non-abstract, non-extern property, the *accessor_body* is one of:

- a *block* that specifies the statements to be executed when the corresponding accessor is invoked
- an expression body, which consists of `=>` followed by an *expression* and a semicolon, and denotes a single expression to be executed when the corresponding accessor is invoked
```

Comment (jskeet) [reply standard/classes.md]: If everyone else is happy with this change, I'd do the same thing for constructor, static constructor and finalizer declarations.

File: standard/classes.md [hunk 2]
```diff
@@ -4413,7 +4418,12 @@ Each of the types referenced in the *formal_parameter_list* of an instance const
...
+When a constructor declaration includes an `extern` modifier, the constructor is said to be an ***external constructor***. Because an external constructor declaration provides no actual implementation, its *constructor_body* consists of a semicolon. For all other constructors, the *constructor_body* consists of either
+
+- a *block*, which specifies the statements to initialize a new instance of the class.
+- or an expression body, which consists of `=>` followed by an *expression* and a semicolon, and denotes a single expression to initialize a new instance of the class.
+
+A *constructor_body* that is a *block* or expression body corresponds exactly to the *block* of an instance method with a `void` return type ([§14.6.11](classes.md#14611-method-body)).
```
Comment (jskeet) [inline standard/classes.md:4435]: This was effectively present before, but I'm not entirely sure what means by "corresponds exactly to" here... should we clarify it, while we're reworking this bit? (I suspect it's in terms of what statements or expressions are valid...)

File: standard/classes.md (hunk 1 above)
Comment (jskeet) [reply standard/classes.md]: Alternatively, we could avoid the bullet points and put it into prose. Here's the equivalent for methods:

> The method_body is either a semicolon, a block body or an expression body. A block body consists of a block, which specifies the statements to execute when the method is invoked. An expression body consists of =>, followed by a null_conditional_invocation_expression or expression, followed by a semicolon, and denotes a single expression to perform when the method is invoked.

**Review by jskeet — COMMENTED** [review]
Basically fine, but a few nits and presentation aspects to discuss.

File: standard/classes.md (hunk 2 above)
Comment (jskeet) [reply standard/classes.md:4435]: Given that this isn't specific to expression-bodied members, I'll raise a separate issue about that.

Comment (jskeet) [reply standard/classes.md:4435]: (Also, Eric noted that this isn't the same meaning of "corresponds" as "A get accessor corresponds to a parameterless method...")

File: standard/classes.md (hunk 1 above)
Comment (jskeet) [reply standard/classes.md]: Having bullets would avoid a massive run-on sentence. Perhaps just change the bulleted list a little.

Jon to try to find other examples and be consistent.

**Comment by jskeet** [comment]
Next steps for Jon:

- File a new issue for the "corresponds exactly to"
- Adjust the bulleted list
- Send out for review

**Review by jskeet — APPROVED** [review]
I've tweaked the list wording - I'm now happy for this to be reviewed more widely.

**Comment by jskeet** [comment]
@RexJaeschke: I note this branch now has conflicts with draft-v7. Would you like me to try to resolve them?

**Comment by RexJaeschke** [comment]
@jskeet . Yes please!

**Comment by jskeet** [comment]
> @jskeet . Yes please!

Righto - I'll do that on Monday.

**Review by gafter — APPROVED** [review]

File: standard/classes.md (hunk 2 above)
Comment (jskeet) [reply standard/classes.md:4435]: Raised #577.

**Comment by jskeet** [comment]
Rebased and squashed to a single commit with a more descriptive message

**Comment by jskeet** [comment]
Ping @BillWagner, @ericlippert, @MadsTorgersen, @Nigel-Ecma for a final review - I'd like to merge soon if possible.

**Review by BillWagner — APPROVED** [review]
This LGTM.

Lets :shipit:

**Pull request merged**